test/widgets/compose_box_test.dart

```diff
@@ -1354,4 +1400,462 @@ void main() {
+      _EditInteractionStart start = _EditInteractionStart.actionSheet,
+      bool? duringFetchRawContentRequest,
+    }) {
+      final description = StringBuffer()..write('tap Cancel ');
+      switch (start) {
+        case _EditInteractionStart.actionSheet:
+          assert(duringFetchRawContentRequest != null);
+          description
+            ..write(duringFetchRawContentRequest! ? 'during ' : 'after ')
+            ..write('fetch-raw-content request: ');
+        case _EditInteractionStart.restoreFailedEdit:
+          assert(duringFetchRawContentRequest == null);
+          description.write('when editing from a restored failed edit: ');
+      }
+      description.write('$narrow');
+      testWidgets(description.toString(), (tester) async {
+        await prepareEditMessage(tester, narrow: narrow);
+        checkNotInEditingMode(tester, narrow: narrow);
+
+        final messageId = msgIdInNarrow(narrow);
+        switch (start) {
+          case _EditInteractionStart.actionSheet:
+            await startInteractionFromActionSheet(tester,
+              messageId: messageId, delay: Duration(seconds: 5));
+            await checkAwaitingRawMessageContent(tester);
+            await tester.pump(duringFetchRawContentRequest!
+              ? Duration(milliseconds: 500)
+              : Duration(seconds: 5));
+          case _EditInteractionStart.restoreFailedEdit:
+            await startInteractionFromRestoreFailedEdit(tester,
+              messageId: messageId,
+              newContent: 'bar');
+            checkContentInputValue(tester, 'bar');
+        }
+
+        await tester.tap(find.widgetWithText(ZulipWebUiKitButton, 'Cancel'));
+        await tester.pump();
+        checkNotInEditingMode(tester, narrow: narrow);
+
+        // We've canceled the previous edit session, so we should be able to
+        // do a new edit-message session…
+        await startInteractionFromActionSheet(tester,
+          messageId: messageId, originalRawContent: 'foo');
+        await checkAwaitingRawMessageContent(tester);
+        await tester.pump(Duration(seconds: 1)); // fetch-raw-content request
+        checkContentInputValue(tester, 'foo');
+        await enterContent(tester, 'qwerty');
+        connection.prepare(json: UpdateMessageResult().toJson());
+        await tester.tap(find.widgetWithText(ZulipWebUiKitButton, 'Save'));
+        checkRequest(messageId, prevContent: 'foo', content: 'qwerty');
+        await tester.pump(Duration.zero);
+        checkNotInEditingMode(tester, narrow: narrow);
+
+        // …or send a new message.
+        connection.prepare(json: {}); // for typing-start request
+        connection.prepare(json: {}); // for typing-stop request
+        await enterContent(tester, 'new message to send');
+        state.controller.contentFocusNode.unfocus();
+        await tester.pump();
+        check(connection.takeRequests()).deepEquals([
+          (Subject<Object?> it) => it.isA<http.Request>()
```

Comment:
nit: this sort of `deepEquals` argument is a bit shorter using `Condition`:
```suggestion
        check(connection.takeRequests()).deepEquals(<Condition<Object?>>[
          (it) => it.isA<http.Request>()
```